### PR TITLE
Synced slider with frame index rather than active keyframe

### PR DESCRIPTION
**Pull request by Fifourche**
A PR to link the `AnimationSlider` with the frame index being set.

Before it was synced with the change of `key_frames.selection.active`, and now with the signal of an `Emitter` : `frame_index`, added to `Animation`. This way, if we use `set_movie_frame_index`, it'll put the slider at the frame's position and not at the closest key frame.

For some context, the idea I had was to sync the slider with a frame_index signal to prepare the way for a "previewer" capability. The preview would be launched via a method of `Animation`, or a button next to the slider.

### CODE REVIEW

File: napari_animation/frame_sequence.py [hunk 1]
```diff
@@ -57,16 +57,20 @@ def _rebuild_frame_index(self, event=None):
...
-        f = 0
-        for kf0, kf1 in pairwise(self._key_frames):
-            for s in range(kf1.steps):
-                fraction = s / kf1.steps
-                self._frame_index[f] = (kf0, kf1, fraction)
-                f += 1
+        else:
+            f = 0
+            if len(self._key_frames) == 1:
+                kf1 = self._key_frames[0]
```
Comment (alisterburt) [inline napari_animation/frame_sequence.py:67]: I don't really understand the purpose of this little section, could you explain it?

**Review by alisterburt — CHANGES_REQUESTED** [review]
hey @Fifourche - I love the idea but am not 100% on the implementation, especially the generation of the whole list to obtain the required index...

Could you also explain the extra stuff you added into the cache? 😄

File: napari_animation/_qt/animation_widget.py [hunk 2]
```diff
@@ -126,27 +129,21 @@ def _on_keyframes_changed(self, event=None):
...
-            self.animationSlider.setValue(frame_index)
-            self.animationSlider.blockSignals(False)
+    def _on_frame_index_changed(self, event=None):
+        frame_index = event.value
+        self.animationSlider.blockSignals(True)
+        self.animationSlider.setValue(frame_index)
+        self.animationSlider.blockSignals(False)
 
+    def _on_active_keyframe_changed(self, event):
+        active_keyframe = event.value
```
[the remark this answers is not in the document]
Comment (Fifourche) [reply napari_animation/_qt/animation_widget.py:139]: Yes and yes ! will do :)

File: napari_animation/frame_sequence.py (hunk 1 above)
Comment (Fifourche) [reply napari_animation/frame_sequence.py:67]: Previously, if there was only one keyframe, it was not added as a frame and _frames was empty. I just added this here so _frames would be empty only when there are no keyframes at all.

**Comment by alisterburt** [comment]
Okay cool! Maybe we could refactor this logic into a ‘has_frames’ properly rather than adding it here?

**Comment by Fifourche** [comment]
Wasn't sure of what you meant, as it's mostly the number of key frames that is needed there. Would you still go for an attribute for this ?

> Okay cool! Maybe we could refactor this logic into a ‘has_frames’ properly rather than adding it here?
> […](#)
> On 1 Jun 2021, at 12:19, Fifourche ***@***.***> wrote: ﻿ @Fifourche commented on this pull request. In napari_animation/frame_sequence.py: > self.events.n_frames(value=len(self)) return - f = 0 - for kf0, kf1 in pairwise(self._key_frames): - for s in range(kf1.steps): - fraction = s / kf1.steps - self._frame_index[f] = (kf0, kf1, fraction) - f += 1 + else: + f = 0 + if len(self._key_frames) == 1: + kf1 = self._key_frames[0] Previously, if there was only one keyframe, it was not added as a frame and _frames was empty. I just added this here so _frames would be empty only when there are no keyframes at all. — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

File: napari_animation/frame_sequence.py [hunk 3]
```diff
@@ -57,16 +57,20 @@ def _rebuild_frame_index(self, event=None):
...
+            else:
+                for kf0, kf1 in pairwise(self._key_frames):
+                    for s in range(kf1.steps):
+                        fraction = s / kf1.steps
+                        self._frame_index[f] = (kf0, kf1, fraction)
+                        f += 1
```
[the remark this answers is not in the document]
Comment (Fifourche) [reply napari_animation/frame_sequence.py]: I still kept an `elif`, for probably a better readability !

File: napari_animation/animation.py [hunk 4]
```diff
@@ -208,6 +230,12 @@ def animate(
         if not save_as_folder:
             writer.close()
 
+    def _keyframe_frame_index(self, keyframe):
+        n_frames = len(self._frames)
+        kf1_list = [self._frames._frame_index[n][0] for n in range(n_frames)]
```
[the remark this answers is not in the document]
Comment (Fifourche) [reply napari_animation/animation.py]: I'm switching back to something I wrote on a previous version. I hope it feels lighter ! :)

**Comment by Fifourche** [comment]
Hey :)
I'm changing a bit this PR to switch the current frame index attribute from `animation` to `FrameSequence` ! In my eyes it makes `FrameSequence` a bit more like an virtual slider, which would be relevant to connect to the GUI slider, and thus prepare the ground for a play button with #100 .

File: napari_animation/animation.py [hunk 5]
```diff
@@ -208,6 +215,15 @@ def animate(
         if not save_as_folder:
             writer.close()
 
+    def _keyframe_frame_index(self, keyframe_index):
+        """Gets the frame index of the keyframe corresponding to keyframe_index."""
+        # Get all steps leading to keyframe.
+        steps_to_keyframe = [
+            kf.steps for kf in self.key_frames[1 : keyframe_index + 1]
+        ]
```
Comment (alisterburt) [inline napari_animation/animation.py:223]: this is looking much nicer! 🙂

File: napari_animation/frame_sequence.py [hunk 6]
```diff
@@ -160,3 +170,17 @@ def iter_frames(
...
+
+    @property
+    def _current_index(self):
+        return self.__current_index
+
+    @_current_index.setter
+    def _current_index(self, frame_index):
+        if frame_index != self._frame_index:
+            self.__current_index = frame_index
+            self.events._current_index(value=frame_index)
```
Comment (alisterburt) [inline napari_animation/frame_sequence.py:186]: I really like having this directly on the `FrameSequence`

**Review by alisterburt — APPROVED** [review]
hey @Fifourche - this is looking really nice now! Thanks for your patience with my reviewing here, let's get this in!

**Pull request merged**